proxy/http/HttpSM.cc

```diff
@@ -1428,34 +1428,19 @@ plugins required to work with sni_routing.
           callout_state = HTTP_API_IN_CALLOUT;
         }
 
-        /* The MUTEX_TRY_LOCK macro was changed so
-           that it can't handle NULL mutex'es.  The plugins
-           can use null mutexes so we have to do this manually.
-           We need to take a smart pointer to the mutex since
-           the plugin could release it's mutex while we're on
-           the callout
-         */
-        bool plugin_lock;
-        Ptr<ProxyMutex> plugin_mutex;
-        if (cur_hook->m_cont->mutex) {
-          plugin_mutex = cur_hook->m_cont->mutex;
-          plugin_lock  = MUTEX_TAKE_TRY_LOCK(cur_hook->m_cont->mutex, mutex->thread_holding);
-
-          if (!plugin_lock) {
-            api_timer = -Thread::get_hrtime_updated();
-            HTTP_SM_SET_DEFAULT_HANDLER(&HttpSM::state_api_callout);
-            ink_assert(pending_action == nullptr);
-            pending_action = mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
-            // Should @a callout_state be reset back to HTTP_API_NO_CALLOUT here? Because the default
-            // handler has been changed the value isn't important to the rest of the state machine
-            // but not resetting means there is no way to reliably detect re-entrance to this state with an
-            // outstanding callout.
-            return 0;
-          }
-        } else {
-          plugin_lock = false;
+        MUTEX_TRY_LOCK(lock, cur_hook->m_cont->mutex, mutex->thread_holding);
+        // Have a mutex but didn't get the lock, reschedule
+        if (!lock.is_locked()) {
+          api_timer = -Thread::get_hrtime_updated();
+          HTTP_SM_SET_DEFAULT_HANDLER(&HttpSM::state_api_callout);
+          ink_assert(pending_action == nullptr);
+          pending_action = mutex->thread_holding->schedule_in(this, HRTIME_MSECONDS(10));
+          // Should @a callout_state be reset back to HTTP_API_NO_CALLOUT here? Because the default
+          // handler has been changed the value isn't important to the rest of the state machine
+          // but not resetting means there is no way to reliably detect re-entrance to this state with an
+          // outstanding callout.
+          return 0;
```

Comment:
@SolidWallOfCode @shinrich 

The plugins which hook on global hooks can use null mutex to handle concurrent request.
Does the plugins are required to have a mutex ?

Reply:
Correct, the continuations are only required to have a mutex if they are using the scheduling APIs.  If cur_hook->m_cont->mutex is null, the changes to MUTEX_TRY_LOCK will just create a lock object that says it is locked.  It will not try to dereference the null mutex.